route_guide/route_guide_server.cc

```diff
@@ -55,166 +66,259 @@ float GetDistance(const Point& start, const Point& end) {
+                        feature_count,
+                        distance,
+                        previous,
+                        start_time](
+                           auto& call,
+                           auto&& point) mutable {
+                   // Check if the request is valid.
+                   if (point) {
+                     point_count++;
+                     if (!GetFeatureName(*point, feature_list_).empty()) {
+                       feature_count++;
+                     }
+                     if (point_count != 1) {
+                       distance += GetDistance(previous, *point);
+                     }
+                     previous = *point;
+                   } else {
+                     // The invalid request indicates the client has finished
+                     // sending all its requests. We can send the final
+                     // response.
+                     system_clock::time_point end_time = system_clock::now();
+                     RouteSummary summary;
+                     summary.set_point_count(point_count);
+                     summary.set_feature_count(feature_count);
+                     summary.set_distance(static_cast<long>(distance));
+                     auto secs = std::chrono::duration_cast<
+                         std::chrono::seconds>(
+                         end_time - start_time);
+                     summary.set_elapsed_time(secs.count());
+                     call.WriteAndFinish(summary, Status::OK);
+                   }
+                 });
+           }))
+        | Loop();
+  }
 
-    status = server->Serve<RouteGuide, Stream<RouteNote>, Stream<RouteNote>>(
-        "RouteChat",
-        [this](auto* call, auto&& note) {
-          if (note) {
-            std::unique_lock<std::mutex> lock(mu_);
-            for (const RouteNote& n : received_notes_) {
-              if (n.location().latitude() == note->location().latitude() &&
-                  n.location().longitude() == note->location().longitude()) {
-                call->Write(n);
-              }
-            }
-            received_notes_.push_back(*note);
-          } else {
-            call->Finish(Status::OK);
-          }
-        },
-        [](auto*, bool) {});
-
-    return status;
+  // ServeRouteChat demonstrates a bidirectional streaming RPC. The client
+  // sends a sequence of RouteNote messages (see protos/route_guide.proto),
+  // which each have a location. The server replies with any RouteNotes that it
+  // has previously received for those locations.
+  auto ServeRouteChat(Server* server) {
+    return server->Accept<
+               RouteGuide,
+               Stream<RouteNote>,
+               Stream<RouteNote>>("RouteChat")
+        | Map(Then([this](auto&& context) {
+             return Server::Handler(std::move(context))
+                 .body([this](auto& call, auto&& note) {
+                   // Check if the request is valid.
+                   if (note) {
```

Comment:
I'd like us to get in the habit of type-safe `if`-statements. As written, `note` could be a pointer, an int, a string, a bool... A small change makes it clear as day:
```suggestion
                   if (note != nullptr) {
```

It's especially helpful here with the `auto&&` typing.

Reply:
Fully approach this idea RJ. One remark: note is `stout::borrowed_ptr<routeguide::RouteNote>` object. That is why the final change will be the following:
    if (note.get() != nullptr) {
      ...
    }

Reply:
Feel free to mark this one resolved when you've fixed it. :)